src/expr/src/scalar/func/impls/float64.rs

```diff
@@ -434,25 +435,35 @@ sqlfunc!(
 
 sqlfunc!(
     #[sqlname = "tots"]
-    fn to_timestamp(f: f64) -> Option<CheckedTimestamp<DateTime<Utc>>> {
-        // TODO(benesch): remove potentially dangerous usage of `as`.
-        #[allow(clippy::as_conversions)]
-        if !f.is_finite() {
-            None
+    fn to_timestamp(f: f64) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
+        if f.is_nan() {
+            Err(EvalError::TimestampCannotBeNan)
+        } else if f.is_infinite() {
+            // TODO(jkosh44) implement infinite timestamps
+            Err(EvalError::TimestampOutOfRange)
         } else {
-            let secs = f.trunc() as i64;
+            let mut secs = f64_to_i64(f.trunc()).ok_or(EvalError::TimestampOutOfRange)?;
             // NOTE(benesch): PostgreSQL has microsecond precision in its timestamps,
             // while chrono has nanosecond precision. While we normally accept
             // nanosecond precision, here we round to the nearest microsecond because
             // f64s lose quite a bit of accuracy in the nanosecond digits when dealing
             // with common Unix timestamp values (> 1 billion).
-            let nanosecs = ((f.fract() * 1_000_000.0).round() as u32) * 1_000;
-            match NaiveDateTime::from_timestamp_opt(secs as i64, nanosecs as u32) {
+            let mut nanosecs = f64_to_i64(((f.fract() * 1_000_000.0).round()) * 1_000.0)
```

Comment:
I think `nanosecs` can be equal to 1 billion here, if `f` is very close to an integer. Say `f` is 1.999999999, then `secs` will be 1 and `nanosecs` will be 1bn -- should it be 2 and 0, instead?

Reply:
Updated to address this and added a test too.